tools/content/unimplemented_features_test.dart

```diff
@@ -0,0 +1,147 @@
+                      'of $totalMessageCount public messages:');
+    }
+    for (final featureName in messageIdsByFeature.keys) {
+      Set<int> messageIds = messageIdsByFeature[featureName]!;
+      int oldestId = messageIds.reduce(min);
+      int newestId = messageIds.reduce(max);
+      outputLines.add('- `$featureName`\n  Oldest message: $oldestId; '
+                      'newest message: $newestId '
+                      '(${messageIds.length}/$failedMessageCount)\n');
+    }
+    outputLines.add('');
+
+    final divider = '\n\n${'=' * 80}\n\n';
+    int unsupportedCounter = 0;
+    for (final MapEntry(key: featureName, value: messageContents) in contentsByFeature.entries) {
+      unsupportedCounter++;
+      if (!_verbose) continue;
+      outputLines.addAll([
+        'Unsupported feature #$unsupportedCounter: $featureName',
+        'message IDs:\n${messageIdsByFeature[featureName]!.join(', ')}',
+        'first 10 examples:\n${messageContents.take(10).join(divider)}',
+        '\n',
+      ]);
+    }
+    check(unsupportedCounter, because: outputLines.join('\n')).equals(0);
+  }
+
+  final corpusFiles = _getCorpusFiles();
+
+  if (corpusFiles.isEmpty) {
+    throw Exception('No corpus found in directory "$_corpusDirPath" to check'
+                    ' for unimplemented features.');
+  }
+
+  group('Check for unimplemented features in', () {
+    for (final file in corpusFiles) {
+      test(file.path, () => checkForUnimplementedFeatureInFile(file));
+    }
+  });
+}
+
+// Determine whether details about all messages with unimplemented features
+// should be printed.
+const bool _verbose = int.fromEnvironment('verbose', defaultValue: 0) != 0;
+
+const String _corpusDirPath = String.fromEnvironment('corpusDir');
+
+Iterable<File> _getCorpusFiles() {
+  final corpusDir = Directory(_corpusDirPath);
+  return corpusDir.existsSync() ? corpusDir.listSync().whereType<File>() : [];
+}
+
+/// Walk the tree looking for unimplemented nodes, and aggregate them by the
+/// category of the unimplemented feature.
+///
+/// This modifies `messageIdsByFeature` and `contentsByFeature` in-place.
+void _walk(int messageId, DiagnosticsNode node, {
+  required Map<String, Set<int>> messageIdsByFeature,
+  required Map<String, List<String>> contentsByFeature,
+}) {
+  final value = node.value;
+  if (value is! UnimplementedNode) {
+    for (final child in node.getChildren()) {
+      _walk(messageId, child,
+        messageIdsByFeature: messageIdsByFeature,
+        contentsByFeature: contentsByFeature);
+    }
+    return;
+  }
+
+  // `featureName` is a prettified identifier used for categorizing
+  // unimplemented features that are likely closely related.
+  final String featureName;
+  final htmlNode = value.debugHtmlNode;
+  if (htmlNode is dom.Element) {
+    if (htmlNode.className.isEmpty) {
+      featureName = '<${htmlNode.localName!}>';
+    } else {
+      featureName = '<${htmlNode.localName!} class="${htmlNode.classes.join(" ")}">';
```

Comment:
```suggestion
      featureName = '<${htmlNode.localName!} class="${htmlNode.className}">';
```

Or is the difference useful?

(`className` is the original data; `classes` is a view on it.)

Reply:
The `classes` view just trims each individual name, I think it is not necessary here.